maro/cli/grass/lib/services/master_agent/local_agent.py

```diff
@@ -0,0 +1,340 @@
+        )
+
+
+class ContainerTrackingAgent(mp.Process):
+    def __init__(self, cluster_name: str, redis_connection, check_interval: int = 60):
+        super().__init__()
+        self.cluster_name = cluster_name
+        self.redis_connection = redis_connection
+        self.check_interval = check_interval
+
+    def run(self):
+        while True:
+            self._check_container_status()
+            time.sleep(self.check_interval)
+
+    def _check_container_status(self):
+        running_jobs = ContainerTrackingAgent.get_running_jobs(
+            self.redis_connection.hgetall(f"{self.cluster_name}:job_details")
+        )
+
+        for job_name, job_detail in running_jobs.items():
+            alive_containers = []
+            container_name_list = job_detail["container_name_list"]
+            for container_name in container_name_list:
+                # Check container status
+                command = f"docker inspect {container_name}"
+                completed_process = subprocess.run(
+                    command,
+                    shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, encoding="utf8"
+                )
+                return_str = completed_process.stdout
+                inspect_details_list = json.loads(return_str)
+
+                # Update container status
+                if not inspect_details_list[0]["State"]["Running"]:
+                    is_continuous = self._container_exit(job_detail, inspect_details_list[0])
+                    if not is_continuous:
+                        break
+                else:
+                    alive_containers.append(container_name)
+
+            if not alive_containers and is_continuous:
+                job_detail["status"] = JobStatus.FINISH
+                self.redis_connection.hset(f"{self.cluster_name}:job_details", job_name, json.dumps(job_detail))
+
+    @staticmethod
+    def get_running_jobs(job_details: dict):
+        name_to_running_job_detail = {}
+        for job_name, job_detail in job_details.items():
+            job_detail = json.loads(job_detail)
+            if job_detail["status"] == JobStatus.RUNNING:
+                name_to_running_job_detail[job_name.decode()] = job_detail
+
+        return name_to_running_job_detail
+
+    def _container_exit(self, job_detail: dict, inspect_details: dict):
+        exit_code = int(inspect_details["State"]["ExitCode"])
+        if exit_code not in ERROR_CODES_FOR_NOT_RESTART_CONTAINER:
+            # Unsuccessfully exited
+            job_detail["status"] = JobStatus.FAILED
+            job_detail["ExitCode"] = exit_code
+            job_detail["Error"] = inspect_details["State"]["Error"]
+            self.redis_connection.hset(f"{self.cluster_name}:job_details", job_detail["name"], json.dumps(job_detail))
+
+            return False
+
+        return True
+
+
+class JobTrackingAgent(mp.Process):
+    def __init__(self, cluster_name: str, redis_connection, check_interval: int = 60):
```

Comment:
int = 60, seems we should trim blanks by following code guidline.

Reply:
It can be set customized in the config of cluster creating.